pkg/vm/stackitem/serialization.go

```diff
@@ -118,9 +126,9 @@ func (w *serContext) serialize(item Item) error {
 				return err
 			}
 		}
-		delete(w.seen, item)
+		w.seen[item] = sliceNoPointer{start, len(w.data)}
```

Comment:
Don't we want to move this line out of switch (i.e. cache other stack item types also, like it is done for JSONisation)?

Reply:
I have checked this now:
```
func BenchmarkEncodeBinaryWithDup(b *testing.B) {
	bi := NewBigInteger(big.NewInt(123))
	arr := NewArray([]Item{
		NewBigInteger(big.NewInt(123)),
		bi,
		NewBool(true),
		bi,
		NewBool(false),
		bi,
	})

	benchmarkEncodeBinary(b, arr)
}

func BenchmarkEncodeBinaryCopy(b *testing.B) {
	arr := NewArray([]Item{
		NewBigInteger(big.NewInt(123)),
		NewBigInteger(big.NewInt(123)),
		NewBool(true),
		NewBigInteger(big.NewInt(123)),
		NewBool(false),
		NewBigInteger(big.NewInt(123)),
	})

	benchmarkEncodeBinary(b, arr)
}
```

Results:
```
name                   old time/op    new time/op    delta
EncodeBinaryWithDup-8     739ns ±14%     771ns ±10%     ~     (p=0.123 n=10+10)
EncodeBinaryCopy-8        743ns ±12%     886ns ± 6%  +19.33%  (p=0.000 n=10+10)

name                   old alloc/op   new alloc/op   delta
EncodeBinaryWithDup-8      480B ± 0%      448B ± 0%   -6.67%  (p=0.000 n=10+10)
EncodeBinaryCopy-8         480B ± 0%      480B ± 0%     ~     (all equal)

name                   old allocs/op  new allocs/op  delta
EncodeBinaryWithDup-8      10.0 ± 0%       8.0 ± 0%  -20.00%  (p=0.000 n=10+10)
EncodeBinaryCopy-8         10.0 ± 0%      10.0 ± 0%     ~     (all equal)
```

Reply:
So I think it doesn't worth the effort (I don't expect a lot of shared references for primitive items).
But we can behave the same way in `ToJSON` then.

Reply:
Fixed `ToJSON`.